primitives/src/crh/poseidon/updatable/mod.rs

```diff
@@ -0,0 +1,217 @@
+        PoseidonHash::<F, P>::poseidon_perm(&mut state);
+        state[0]
+    }
+
+}
+
+impl<F, P> UpdatableFieldBasedHash for UpdatablePoseidonHash<F, P>
+    where
+        F: PrimeField + MulShort,
+        P: PoseidonParameters<Fr = F>,
+{
+    type Data = F;
+    type Parameters = P;
+
+    fn update(&mut self, input: &Self::Data) -> &mut Self {
+        self.pending.push(*input);
+        if self.pending.len() == P::R {
+            self.apply_permutation();
+            self.pending.clear();
+        }
+        self
+    }
+
+    fn finalize(&self) -> Self::Data {
+        if !self.pending.is_empty() {
+            self._finalize()
+        } else {
+            self.state[0]
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use algebra::{fields::{mnt4753::Fr as MNT4Fr, mnt6753::Fr as MNT6Fr}, UniformRand, Field};
+    use rand::SeedableRng;
+    use rand_xorshift::XorShiftRng;
+    use super::*;
+
+    type UpdatableMNT4PoseidonHash = UpdatablePoseidonHash<MNT4Fr, MNT4753PoseidonParameters>;
+    type UpdatableMNT6PoseidonHash = UpdatablePoseidonHash<MNT6Fr, MNT6753PoseidonParameters>;
+
+    #[test]
+    fn test_updatable_poseidon_hash_mnt4(){
+        let samples = 100;
+        let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+        // Test consistency with PoseidonHash and UpdatablePoseidonHash
+        for i in 1..=samples {
+            let input = vec![MNT4Fr::rand(&mut rng); i];
+            let hash_output = MNT4PoseidonHash::evaluate(input.as_slice()).unwrap();
+            let updatable_hash_output = {
+                let mut updatable_hash = UpdatableMNT4PoseidonHash::new(None);
+                for input in input.into_iter() {
+                    updatable_hash.update(&input);
+                }
+                updatable_hash.finalize()
+            };
+            assert_eq!(hash_output, updatable_hash_output, "Hashes output with {} inputs must be equal", i);
+        }
+
+        // Test finalize() holding the state and allowing updates in between different calls to it
+        let input = vec![MNT4Fr::rand(&mut rng); 2];
+        let h_out = MNT4PoseidonHash::evaluate(input.as_slice()).unwrap();
+
+        let mut uh =  UpdatableMNT4PoseidonHash::new(None);
+        uh.update(&input[0]);
+        uh.finalize();
+        uh.update(&input[1]);
+        assert_eq!(h_out, uh.finalize());
+
+        //Test finalize() being idempotent
+        assert_eq!(h_out, uh.finalize());
+
+        // Test initializing UpdatablePoseidonHash with personalization is the same as concatenating
+        // to PoseidonHash input the personalization and the padding.
+        let input = MNT4Fr::rand(&mut rng);
+        let samples = 10;
+        for i in 1..=samples{
+            let personalization = vec![MNT4Fr::rand(&mut rng); i];
+
+            let mut hash_input = personalization.clone();
+            let padding = vec![MNT4Fr::zero(); personalization.len() % MNT4753PoseidonParameters::R];
```

Comment:
Shouldn't be MNT4753PoseidonParameters::R - personalization.len() % MNT4753PoseidonParameters::R?